### PR TITLE
fix: treat input spent as non-canonical

### PR DESCRIPTION
**Pull request by boolafish**
### Note
Changes the behavior when input is flagged as spent. Previously, it would be omitted while now it
would be consider non-canonical instead. Omitting would lead to edge case that some input would be locked to the system while it should be able to exit.

Also, this makes changes to how we flag inputs/outputs:
1) flag only when finalizing the exit
2) for canonical mark all inputs spend and exited outputs spend
3) for non-canonical mark only exiting inputs

ref: https://github.com/omisego/plasma-contracts/issues/470

### Side change and refactor
- Add `isInputEmpty`, `isOutputEmpty` functions to `PaymentInFlightExitModelUtils`
- Use `isXXXEmpty` check instead of `amount > 0`
- apply `isXXXEmpty` to piggyback functions to make them fail more explicitly
- fix `PaymentInFlightExitModelUtils` tests that have the for loop written wrong....
- move tests for controller into `controllers/` folder. Too many PaymentXXX tests now.

### TODO
- [x] run docgen

### Test
```
      When any input is spent, given the challenge game result is canonical
        ✓ should be treat as non canonical (140ms)
      When the exit is non canonical, and some inputs/outputs are piggybacked
        ✓ should flag and only flag piggybacked inputs with the same token as spent (136ms)
        ✓ should NOT flag output as spent (145ms)
      When the exit is canonical, and some inputs/outputs are piggybacked
        ✓ should flag ALL inputs with the same token as spent (166ms)
        ✓ should flag and only flag piggybacked output with the same token as spent (169ms)
```

### CODE REVIEW

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 1]
```diff
@@ -193,25 +197,45 @@ library PaymentProcessInFlightExit {
...
     )
         private
     {
-        // We flag _all_ inputs regardless of whether it is piggybacked
-        // If exiting from output, all inputs are considered spent and can only exit from output in future.
-        // If exiting from input, to keep things simple, all users must piggyback the input at the same time,
-        // instead of re-starting the IFE and then re-piggyback their input.
         uint256 inputNumOfTheToken;
         for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
-            if (exit.inputs[i].token == token && exit.inputs[i].amount > 0) {
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: does not seem necessary to check whether amount exists or not? We set data for all inputs and outputs in start IFE no matter the amount value.

Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: ah....okay. So we use fixed array in IFE exit data so it is actually needed.

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 2]
```diff
@@ -429,17 +404,46 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
...
 
-        describe('When the exit is non canonical, and inputs are piggybacked', () => {
+        describe('When any input is spent, given the challenge game result is canonical', () => {
+            beforeEach(async () => {
+                const exit = await buildInFlightExitData();
+                this.dummyExitId = 666;
+                exit.isCanonical = true;
+                await this.exitGame.setInFlightExit(this.dummyExitId, exit);
+
+                // flags the first input and piggyback the two ETH inputs (first and second)
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: piggyback -> piggybacks

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 3]
```diff
@@ -429,17 +404,46 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
...
                 this.dummyExitId = 666;
                 this.exit = await buildInFlightExitData();
                 this.exit.isCanonical = false;
 
-                // piggybacks all three inputs
-                this.exit.exitMap = (2 ** 0) | (2 ** 1) | (2 ** 2);
-
                 await this.exitGame.setInFlightExit(this.dummyExitId, this.exit);
 
+                // piggybackes the first input for ETH and third for ERC20
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: Duplicated line.

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 4]
```diff
@@ -488,6 +482,24 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
                 expect(postBalance).to.be.bignumber.equal(expectedBalance);
             });
 
+            it('should flag and only flag piggybacked inputs with the same token as spent', async () => {
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: flag and only flag -> flag only

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 5]
```diff
@@ -429,17 +404,46 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
...
+                exit.isCanonical = true;
+                await this.exitGame.setInFlightExit(this.dummyExitId, exit);
+
+                // flags the first input and piggyback the two ETH inputs (first and second)
+                await this.exitGame.proxyFlagOutputSpent(TEST_OUTPUT_ID_FOR_INPUT_1);
+                await this.exitGame.setInFlightExitInputPiggybacked(this.dummyExitId, 0);
+                await this.exitGame.setInFlightExitInputPiggybacked(this.dummyExitId, 1);
+            });
+
+            it('should be treat as non canonical', async () => {
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: treat -> treated

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 6]
```diff
@@ -429,17 +404,46 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
             });
         });
 
-        describe('When the exit is non canonical, and inputs are piggybacked', () => {
+        describe('When any input is spent, given the challenge game result is canonical', () => {
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: I understand that's the scenario described in the issue. Right?

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 7]
```diff
@@ -429,17 +404,46 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
...
+                // flags the first input and piggyback the two ETH inputs (first and second)
+                await this.exitGame.proxyFlagOutputSpent(TEST_OUTPUT_ID_FOR_INPUT_1);
+                await this.exitGame.setInFlightExitInputPiggybacked(this.dummyExitId, 0);
+                await this.exitGame.setInFlightExitInputPiggybacked(this.dummyExitId, 1);
+            });
+
+            it('should be treat as non canonical', async () => {
+                const { logs } = await this.exitGame.processExit(this.dummyExitId, VAULT_ID.ETH, ETH);
+                const inputIndexForInput2 = 1;
+                await expectEvent.inLogs(
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: It would be good, if we could check that there was no event for inputIndex 0.

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js [hunk 8]
```diff
@@ -354,14 +339,18 @@ contract('PaymentInFlightExitRouter', ([_, ifeBondOwner, inputOwner1, inputOwner
 
                 this.dummyExitId = 666;
                 this.exit = await buildInFlightExitData();
+                await this.exitGame.setInFlightExit(this.dummyExitId, this.exit);
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: Everywhere in this file we use dummyExitId with value 666. How about, we put it into a constant?

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 9]
```diff
@@ -133,14 +138,14 @@ library PaymentProcessInFlightExit {
...
-            if (exit.inputs[i].token == token) {
+            if (exit.inputs[i].token == token && exit.inputs[i].amount > 0) {
                 inputNumOfTheToken++;
             }
         }
         bytes32[] memory outputIdsOfInputs = new bytes32[](inputNumOfTheToken);
         uint sameTokenIndex = 0;
         for (uint i = 0; i < MAX_INPUT_NUM; i++) {
-            if (exit.inputs[i].token == token) {
+            if (exit.inputs[i].token == token && exit.inputs[i].amount > 0) {
```
Comment (pgebal) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: Why was that added check necessary?

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 10]
```diff
@@ -158,9 +163,8 @@ library PaymentProcessInFlightExit {
         pure
         returns (bool)
     {
-        // The check to determine whether input is spent, performed in 'isAnyInputSpent'
-        // For this reason, no need to check again here.
         return withdrawal.token == token &&
+                exit.isInputPiggybacked(index) &&
```
[the remark this answers is not in the document]
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: aww, I need to add check on the `isOutputSpent` (I think) as we are not omitting the process anymore but I copied the wrong line @@

[the remark this answers is not in the document]
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: also, test added for the case that if already flagged, would not withdraw

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js (hunk 6 above)
Comment (boolafish) [reply plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: yup. Also some other tests as put in the description.

File: plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js (hunk 7 above)
Comment (boolafish) [reply plasma_framework/test/src/exits/payment/PaymentProcessInFlightExit.test.js]: added the test under non-canonical test cases (describe) instead of here.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol (hunk 9 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: I want to be consistent. We flag only when the input is not empty however we check inputs even if it is empty. It only works because empty input --> outputId == 0 --> the flag of that would always be false. I think it is better to not check the flag for that.

Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: Also, I abstract the check of `amount > 0` to be more explicit checking whether it is an empty withdraw data and have a function in library for that.

File: plasma_framework/contracts/mocks/exits/payment/PaymentInFlightExitModelUtilsMock.sol [hunk 11]
```diff
@@ -13,6 +13,43 @@ contract PaymentInFlightExitModelUtilsMock {
...
+        } else {
+            revert("target should be either inputs or outputs");
+        }
+    }
+
+    function stringEquals(string memory a, string memory b) private pure returns (bool) {
+        return keccak256(abi.encodePacked(a)) == keccak256(abi.encodePacked(b));
+    }
+
+
```
Comment (pgebal) [inline plasma_framework/contracts/mocks/exits/payment/PaymentInFlightExitModelUtilsMock.sol]: We usually use one empty line for spacing. Here we have 2.

File: plasma_framework/test/src/exits/payment/PaymentInFlightExitModelUtils.test.js [hunk 12]
```diff
@@ -3,22 +3,226 @@
...
+            for (let i = 0; i < MAX_INPUTS; i++) {
+                const actual = await modelUtils.isInputEmpty(i);
+                expect(actual).to.be.true;
+            }
+        });
+
+        it('returns false when the outputId is not empty', async () => {
+            const modelUtils = await PaymentInFlightExitModelUtils.new(0, 0);
+
+            const dataWithNonEmptyOutputId = {
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentInFlightExitModelUtils.test.js]: I would make a constant `emptyWithdrawData` on a higher scope and in each test I would just update one value in it. That would save some lines of code.
